src/scattdata_header.F90

```diff
@@ -482,8 +637,85 @@ subroutine scattdatatabular_sample(this, gin, gout, mu, wgt)
+    wgt = wgt * this % mult(gin) % data(gout)
 
   end subroutine scattdatatabular_sample
 
+!===============================================================================
+! SCATTDATA*_GET_MATRIX Reproduces the original scattering matrix (densely)
+! using ScattData's information of fmu/dist, energy, and scattxs
+!===============================================================================
+
+    pure function scattdata_get_matrix(this, req_order) result(matrix)
+      class(ScattData), intent(in) :: this          ! Scattering Object to work with
+      integer, intent(in)          :: req_order     ! Requested order of matrix
+      real(8), allocatable         :: matrix(:,:,:) ! Resultant matrix just built
+
+      integer :: order, groups, gin, gout
+
+      groups = size(this % energy)
+      order = min(req_order,size(this % dist(1) % data(:,1)))
+
+      allocate(matrix(order,groups,groups))
+      ! Initialize to 0; this way the zero entries in the dense matrix dont
+      ! need to be explicitly set, requiring a significant increase in the
+      ! lines of code.
+      matrix = ZERO
+      do gin = 1, groups
+        do gout = this % gmin(gin), this % gmax(gin)
+          matrix(:,gout,gin) = this % scattxs(gin) * &
+               this % energy(gin) % data(gout) * &
+               this % dist(gin) % data(1:order,gout)
+        end do
+      end do
+    end function scattdata_get_matrix
+
+    pure function scattdatahistogram_get_matrix(this, req_order) result(matrix)
+      class(ScattDataHistogram), intent(in) :: this          ! Scattering Object to work with
+      integer, intent(in)                 :: req_order     ! Requested order of matrix
+      real(8), allocatable                :: matrix(:,:,:) ! Resultant matrix just built
+
+      integer :: order, groups, gin, gout
+
+      groups = size(this % energy)
+      order = min(req_order,size(this % dist(1) % data(:,1)))
+
+      allocate(matrix(order,groups,groups))
+      ! Initialize to 0; this way the zero entries in the dense matrix dont
+      ! need to be explicitly set, requiring a significant increase in the
+      ! lines of code.
+      matrix = ZERO
+      do gin = 1, groups
+        do gout = this % gmin(gin), this % gmax(gin)
+          matrix(:,gout,gin) = this % scattxs(gin) * &
+               this % energy(gin) % data(gout) * &
+               this % fmu(gin) % data(1:order,gout)
+        end do
+      end do
+    end function scattdatahistogram_get_matrix
+
+    pure function scattdatatabular_get_matrix(this, req_order) result(matrix)
+      class(ScattDataTabular), intent(in) :: this          ! Scattering Object to work with
+      integer, intent(in)                 :: req_order     ! Requested order of matrix
+      real(8), allocatable                :: matrix(:,:,:) ! Resultant matrix just built
+
+      integer :: order, groups, gin, gout
+
+      groups = size(this % energy)
+      order = min(req_order,size(this % dist(1) % data(:,1)))
```

Comment:
Yeah, and that would fix a bug I just realized in that the array may not be indexed by 1.